### ISSUE TITLE
Incorrect imports added for ngInjectableDef with Bazel

### ISSUE THREAD

**Issue by jelbourn** [angular/angular#25644]
### Summary
When building Angular Material with Bazel (including Angular from source), incorrect import statements are being written for `ngInjectableDef` statements.

### Steps to reproduce
1. Clone `https://github.com/jelbourn/material2/tree/build-angular-from-source`
2. Grab a local snapshot of Angular for debugging:
```sh
wget https://github.com/angular/angular/archive/4e36f0cd68dfc5e0b9f53f6555cbc49a38efdcf7.zip -O /tmp/ng.zip
(cd /tmp ; unzip ng.zip)
```
3. `bazel build src/cdk:npm_package`
4. Examine the file:
```
$(bazel info bazel-bin)/src/cdk/npm_package.es6/src/cdk/drag-drop/drag-drop-registry.js
```
5. Note the imports:
```ts
import * as i1 from "@angular/cdk/overlay/public-api.ngfactory";
import * as i2 from "@angular/cdk/bidi/dir-document-token.ngfactory";
```
6. and the `ngInjectableDef` statement:
```ts
DragDropRegistry.ngInjectableDef = i0.defineInjectable({ factory: function DragDropRegistry_Factory() { return new DragDropRegistry(i0.inject(i1.NgZone_64), i0.inject(i2.DOCUMENT_4)); }, token: DragDropRegistry, providedIn: "root" });
```

Both injected tokens (`DOCUMENT` and `NgZone`) are being imported from the wrong place.

cc @IgorMinar

**Comment by alexeagle** [comment]
There are no npm package publishes involved here. It should look like Google3.

**Comment by jelbourn** [comment]
@alexeagle The difference is rollup getting involved here to _create_ an npm package. In google3 it works fine because there's nothing wrong with referencing the ngfactory files.

**Comment by alexeagle** [comment]
So that means that if Angular itself ever had tree-shakable providers, we
would not be able to build Angular's packages with Bazel any more?

**Comment by devversion** [comment]
Just for tracking of what I found so far. Some things are already clear, but I want to reference them with their code parts.

Currently whenever a factory file gets created, external imports needed for the factory file are imported and re-exported as part of the same factory file. 

https://github.com/angular/angular/blob/73dcd72afbd782161c1674f1e647a1b357a78a67/packages/compiler/src/aot/compiler.ts#L557-L562

This is done because the corresponding `.ngsummary` file declares these external imports (e.g. `NgZone` or `DOCUMENT`) as exported from _that_ factory file. 

Now if we run NGC again with the `ngfactory` and `ngsummary` files referenced and `"generateCodeForLibraries": false`, the summary is deserialized and whenever the actual source files for example depend on `NgZone`, the `SummaryResolver` resolves the import expression to the factory re-export.

https://github.com/angular/angular/blob/73dcd72afbd782161c1674f1e647a1b357a78a67/packages/compiler/src/aot/summary_resolver.ts#L86-L89

This could be avoided by:

1. Using `generateCodeForLibraries != false`. This option is set to `false` within Bazel by default. (see [here](https://github.com/angular/angular/blob/69198ba37bc156ca16034357b5be0dfb3c22e719/packages/compiler-cli/src/transformers/program.ts#L676-L681))
2. Or passing through the original `StaticSymbol` for external import expressions.

---
**Approach 2**

Instead of outputting a summary file for a file _X_ that declares external imports as exported within file X, we just pass through the original location of the external symbol.

This would mean that we don't need to re-export symbols that are already exported, and that output files don't incorrectly import symbols through a layer of indirection (e.g. through `factory` file). e.g. the summary looks like that:

```json
{
    "__symbol": 9,
    "name": "NgZone",
    "filePath": "angular_material/external/angular/packages/core/src/zone/ng_zone",
    "importAs": {
        "filePath": "C:/users/paul/_bazel_paul/kn4tsvyh/execroot/angular_material/bazel-out/x64_windows-fastbuild/bin/external/angular/packages/core/src/zone/ng_zone.d.ts",
        "name": "NgZone",
        "members": []
    }
},
```

Previously it looked like that:

```json
    {
      "__symbol": 9,
      "name": "NgZone",
      "filePath": "angular_material/external/angular/packages/core/src/zone/ng_zone",
      "importAs": "NgZone_9"
    },
```

This change would sound pretty reasonable to me, and could be implemented easily without big efforts since there already is a similar logic: [see here](https://github.com/angular/angular/blob/master/packages/compiler/src/aot/summary_serializer.ts#L147-L153).

---

**tl;dr**: There seem to be ways to fix this issue quite easily. Approach (1) would potentially require a lot of refactoring of the Angular Bazel rules, and approach (2) might be a bit risky, but sounds more reasonable and should be quite easy to implement.

**Comment by alxhub** [comment]
@devversion I think your understanding of summaries and how they work now matches (if not exceeds) my own.

One issue that I can think of:

> This would mean that we don't need to re-export symbols that are already exported, and that output files don't incorrectly import symbols through a layer of indirection (e.g. through factory file).

This actually happens for a specific reason. In google3 we have this concept of "strict deps" - if you import symbol X from module M, then at the BUILD level you have to have a direct dependency on the package that builds M. If you don't, the build system gives you an error.

Angular, as it turns out, doesn't play nicely with this system in certain cases. For example, BrowserModule (in @angular/platform-browser) exports CommonModule (from @angular/common), meaning users of BrowserModule can use NgIf and NgFor in their applications.

If the user writes a template with an NgIf, the NgFactory for that template will need to instantiate NgIf somehow, which means it must import it from somewhere. But where? If it imports it from @angular/common, then the generated code is _creating_ a dependency on @angular/common where there wasn't one before - the user didn't need to know anything about @angular/common to get to this point. If we generated that import, we could fail strict deps checking if the user didn't actually configure that dependency. This is actually how Angular _should_ work, and how Ivy _does_ work.

Instead, we did this crazy solution where factories/summaries re-export types, and we import things via these convoluted re-exports rather than directly from the source. That way, the user's factory from above can import NgIf via BrowserModule and its factory, and there's no strict deps violation - the user doesn't need a hard dependency on @angular/common for this to work.

Maybe the right answer here is for Blaze and Bazel to behave differently, with Blaze continuing to write and consume the re-exports, and Bazel having more simplified metadata/summaries with direct references? It's stretching my understanding of how this whole system is put together, but that seems like it could work to me.

**Comment by devversion** [comment]
@alxhub That's reasonable. I'm just not too sure how Blaze currently invokes NGC, and how I can implement different behavior in that case. I assume G3 also uses the `ng_module` rules right?

**Comment by alxhub** [comment]
@devversion correct. In https://github.com/angular/angular/pull/28523 I introduce a flag to the compiler to switch behavior between Bazel/Blaze, you could follow this same model.
